docs/docs/how-to/plugins-and-themes/adding-gatsby-image-support.md

```diff
@@ -0,0 +1,261 @@
+
+There are three steps to create a custom image component:
+
+1. [Create your URL builder function](#create-your-url-builder-function)
+2. [Create your image data function](#create-your-image-data-function)
+3. [Create your wrapper component](#create-your-wrapper-component) (optional)
+
+### Create your URL builder function
+
+This is similar to the `generateImageSource` described above, but just returns a URL string. This is an example for the same image host:
+
+```js
+function urlBuilder({ baseUrl, width, height, format, options }) {
+  return `https://myexampleimagehost.com/${baseURL}?w=${width}&h=${height}&fmt=${format}&q=${options.quality}`
+}
+```
+
+If your host supports it, we recommend using auto-format to deliver next-generation image formats to supported browsers. In this case, ignore the `format` option.
+
+### Create your image data function
+
+This is a similar to the image resolver described above. However because it executes in the browser, the functions that you use should be fast, and synchronous, and you can't use node APIs. You will not be downloading and generating base64 placeholders, for example, or calculating dominant colors. If you have these values pre-calculated then you can pass these in and use them, but they need to be available in the props that you pass to the function at runtime.
+
+The function should accept the props that will be passed into your component, and at a minimum needs to take the props required by the `getImageData` helper function from `gatsby-plugin-image`. Here is an example for an image host:
+
+```js
+import { getImageData } from "gatsby-plugin-image"
+
+export function getExampleImageData({ image, ...props }) {
+  return getImageData({
+    baseUrl: image.url,
+    sourceWidth: image.width,
+    sourceHeight: image.height,
+    urlBuilder,
+    pluginName: "gatsby-source-example",
+    // If your host supports auto-format/content negotiation, pass this as the formats array
+    formats: ["auto"],
+    ...props,
+  })
+}
+```
+
+You can export this function as a public API, and users can use the function to generate data to pass to `GatsbyImage`:
+
+```jsx
+// This might come from an API at runtime
+const image = {
+  url: "kitten.jpg",
+  width: 800,
+  height: 600,
+}
+const imageData = getExampleImageData({ image, layout: "fixed", width: 400 })
+return <GatsbyImage image={imageData} alt="Kitten" />
+```
+
+### Create your wrapper component
+
+This stage is optional: you may prefer to just share the image data function and let your users pass the result to `<GatsbyImage>`, as shown above. However the developer experience is better with a custom image component.
+
+The component should accept the same props as your image data function, and also all of the props for `<GatsbyImage>` too, which it can pass down to that component. Here's how you might type the props in TypeScript:
```

Comment:
```suggestion
The component should accept the same props as your image data function, as well as all of the props for `<GatsbyImage>` which it can pass down to that component. Here's how you might type the props in TypeScript:
```